### PR TITLE
Added tests to the project

### LINKED ISSUE #141: Add tests to the project
This project lacks of tests, to check if changes to the Dockerfile introduce any bug.

### PR DESCRIPTION
**Pull request by krufab**
Added tests to the project, as per issue #141 and [issue 100](https://github.com/jenkinsci/docker-slave/issues/100) in [docker-slave](https://github.com/jenkinsci/docker-slave)

Closes #141

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Thanks! We will need to resubmit the PR to check whether the Jenkinsfile really works on the CI instance, but the code looks good overall

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,13 @@
 ROOT:=$(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
-IMAGE_NAME:=jenkins4eval/jnlp-slave:test
+IMAGE_NAME:=jenkins4eval/jnlp-slave
 
 build:
-	docker build -t ${IMAGE_NAME} .
-	docker build -t ${IMAGE_NAME}-alpine -f Dockerfile-alpine .
-	docker build -t ${IMAGE_NAME}-jdk11 -f Dockerfile-jdk11 . 
+	docker build -t ${IMAGE_NAME}:latest .
```
Comment (slide) [inline Makefile]: Ideally, I think this should actually be split like we have on the master docker image Makefile. Being able to build a single platform is nice.

File: tests/tests.bats [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  run docker exec netcat-helper /bin/sh -c "timeout 10s nc -l 5000"
+
+  # The GET request ends with a '\r'
+  [ $'GET /tcpSlaveAgentListener/ HTTP/1.1\r' = "${lines[0]}" ]
+}
+
+@test "[${FLAVOR}] use build args correctly" {
+  cd "${BATS_TEST_DIRNAME}"/.. || false
+
+	local ARG_TEST_VERSION
```
Comment (slide) [inline tests/tests.bats]: Whitespace difference?

Comment (krufab) [reply tests/tests.bats]: Hidden tabs..

File: Makefile (hunk 1 above)
Comment (krufab) [reply Makefile]: Updated.

**Comment by krufab** [comment]
Please note that the tests are not being executed as the modified Jenkinsfile isn't processed. Besides, I guess bats won't be available on the machine..

File: Jenkinsfile [hunk 3]
```diff
@@ -57,7 +57,9 @@ pipeline {
                             if(!infra.isTrusted()) {                                
                                 deleteDir()
                                 checkout scm
-                                sh "make build ; docker system prune --force --all"
+                                sh 'make build'
```
Comment (slide) [inline Jenkinsfile]: I'm not trying to pedantic, but you could combine this into a single `sh` call to make it a tiny bit more efficient:

```
sh '''
make build
make test
docker system prune --force --al
```

**Comment by slide** [comment]
I created #143 to test the build

File: Jenkinsfile (hunk 3 above)
Comment (krufab) [reply Jenkinsfile]: Yeah I know. Does it exit if one of the commands fails or should I set -e?

Comment (slide) [reply Jenkinsfile]: Yes, it should. Though we want the prune to always run if possible.

**Comment by krufab** [comment]
Is it ok this PR or is there anything to change?

**Comment by oleg-nenashev** [comment]
Ignoring the CI issues

**Comment by oleg-nenashev** [comment]
@slide @krufab  Looks like CI fails after all

```
[2020-01-29T16:39:16.098Z] + make test

[2020-01-29T16:39:16.098Z] Cloning into 'bats-core'...

[2020-01-29T16:39:16.356Z] HEAD is now at c706d14 Bats 1.1.0

[2020-01-29T16:39:20.528Z] 1..4

[2020-01-29T16:39:20.784Z] ok 1 [alpine] build image

[2020-01-29T16:39:24.964Z] ok 2 [alpine] image has installed jenkins-agent in PATH

[2020-01-29T16:39:39.824Z] ok 3 [alpine] image starts jenkins-agent correctly

[2020-01-29T16:39:58.494Z] ok 4 [alpine] use build args correctly

[2020-01-29T16:39:58.494Z] 1..4

[2020-01-29T16:39:58.494Z] ok 1 [debian] build image

[2020-01-29T16:40:01.027Z] ok 2 [debian] image has installed jenkins-agent in PATH

[2020-01-29T16:40:15.882Z] not ok 3 [debian] image starts jenkins-agent correctly

[2020-01-29T16:40:15.882Z] # (in test file tests/tests.bats, line 59)

[2020-01-29T16:40:15.882Z] #   `[ $'GET /tcpSlaveAgentListener/ HTTP/1.1\r' = "${lines[0]}" ]' failed with status 124

[2020-01-29T16:40:15.882Z] # 1e24fe7e97e28d951d3da2c5e844409ac5330aa373f65d9b7aecc6d2b0c96ef9

[2020-01-29T16:40:15.882Z] # 9242597faa79c429ab5d8ea96714926472644b887b3add40a799a2984432cd68

[2020-01-29T16:40:15.882Z] # /tmp/bats.25649.src: line 59: lines[0]: unbound variable

[2020-01-29T16:41:02.522Z] ok 4 [debian] use build args correctly

[2020-01-29T16:41:02.522Z] Makefile:33: recipe for target 'test-debian' failed

[2020-01-29T16:41:02.522Z] make: *** [test-debian] Error 1
```

**Comment by krufab** [comment]
I'll check it in few hours

**Comment by krufab** [comment]
The failure is probably due to the fact that it took too much time to start the jenkins jnlp agent image. Test 3 completes successfully for alpine and it's the same code for the debian one.
The test firstly spins off a docker jnlp agent in a detached mode, then starts a netcat container listening on port 5000 which waits for 10s to have the agent connect. After 10s the netcat container shuts down and returns the header read (if any). If the jnlp container doesn't boot it that time, the test fails.

To fix it, we can extend the timeout to i.e. 30s. I can create a branch to patch master or I can resolve the issue in the next full change branch that I will submit soon.

**Comment by oleg-nenashev** [comment]
Small change for this issue would be great

**Comment by krufab** [comment]
Done in #148

**Pull request merged**